### PR TITLE
Add cache

### PR DESCRIPTION
**Pull request by ozgunozerk**
Yanking version to 0.4.0 since this is a breaking change to summary, due to a new field added.

Changes:
- summary -> this is now a cache
- summary_file -> this is the file for persistent storage for the summary info

Logic (very high level):
- on start, read the summary_file and initialize the cache
- on events, update the cache accordingly
- at every 100 block, write cache to persistent storage

### CODE REVIEW

File: src/commands/farm.rs [hunk 1]
```diff
@@ -239,30 +239,28 @@ async fn subscribe_to_plotting_progress(
...
+            .get_events(Some(block_hash))
             .await
             .into_eyre()
-            .context("couldn't get events from node")?;
-
-        for event in events {
-            // subscription is active when plotting is started, only print out rewards after
-            // plotting finishes to not corrup the progress bars
+            .context("couldn't get event from node")?
+        {
```
Comment (i1i1) [inline src/commands/farm.rs:263]: I would parallelize it using [async_stream](https://docs.rs/async-stream/latest/async_stream/):
```rust
    fn missed_blocks_stream(
        node: std::sync::Arc<Node>,
        mut last_block_num: subspace_sdk::node::BlockNumber,
    ) -> impl Stream<Item = anyhow::Result<subspace_sdk::node::BlockNumber>> {
        async_stream::try_stream! {
            loop {
                let last_retrieved_num = node.get_info().await?.finalized_block.1;

                if last_block_num == last_retrieved_num {
                    break;
                }

                while last_block_num < last_retrieved_num {
                    yield last_block_num;
                    last_block_num += 1;
                }
            }
        }
    }

    let concurency_parameter = todo!();

    let node = std::sync::Arc::new(todo!());
    let s = missed_blocks_stream(std::sync::Arc::clone(&node), todo!());

    futures::pin_mut!(s);

    let missed_rewards: u128 = s
        .map(|block_number| {
            Ok(node.get_events(Some(
                node.block_hash(block_number?)?
                    .expect("Should be always present if no pruning enabled"),
            )))
        })
        .try_buffer_unordered(concurency_parameter)
        .map_ok(|events| futures::stream::iter(events).map(anyhow::Result::Ok))
        .try_flatten()
        .map_ok(|event| match event {
            Event::Rewards(
                RewardsEvent::VoteReward { voter: author, reward }
                | RewardsEvent::BlockReward { block_author: author, reward },
            ) if author == reward_address.into() => reward,
            _ => 0,
        })
        .try_fold(0u128, |acc, rewards| {
            futures::future::ready::<anyhow::Result<_>>(Ok(acc + rewards))
        })
        .await?;
```

Comment (i1i1) [reply src/commands/farm.rs:263]: Sorry, missed part with farmed block count here, but I think you get an idea.

File: src/summary.rs [hunk 2]
```diff
@@ -40,39 +40,114 @@ impl std::fmt::Display for Rewards {
...
-/// command
-#[derive(Deserialize, Serialize, Debug)]
+/// struct for flexibly updating the fields of the summary
+#[derive(Default, Debug)]
+pub(crate) struct SummaryUpdateFields {
+    pub(crate) is_plotting_finished: bool,
+    pub(crate) is_new_block_farmed: bool,
+    pub(crate) is_new_vote: bool,
+    pub(crate) maybe_new_reward: Option<Rewards>,
+    pub(crate) maybe_last_block: Option<u64>,
```
Comment (i1i1) [inline src/summary.rs:50]: I think you should cache these 4 things:
- last seen finalized block
- total votes
- total farmed blocks
- total rewards

File: src/commands/farm.rs (hunk 1 above)
Comment (i1i1) [reply src/commands/farm.rs:263]: You can update summary somewhere in between here

File: src/summary.rs (hunk 2 above)
Comment (ozgunozerk) [reply src/summary.rs:50]: but I do require `is_plotting_finished` as well

Comment (i1i1) [inline src/summary.rs:50]:
```suggestion
    pub(crate) is_plotting_finished: bool,
    pub(crate) new_block_farmed: u32,
    pub(crate) new_vote: bool,
    pub(crate) new_rewards: Rewards,
    pub(crate) maybe_last_block: Option<u64>,
```

File: src/summary.rs [hunk 3]
```diff
@@ -40,39 +40,114 @@ impl std::fmt::Display for Rewards {
...
+impl SummaryFile {
     /// creates new summary file pointer
     ///
     /// if user_space_pledged is provided, it creates a new summary file
     /// else, it tries to read the existing summary file
     #[instrument]
-    pub(crate) async fn new(user_space_pledged: Option<ByteSize>) -> Result<SummaryFilePointer> {
+    pub(crate) async fn new(user_space_pledged: Option<ByteSize>) -> Result<SummaryFile> {
         let summary_path = summary_path();
         let summary_dir = dirs::data_local_dir()
```
Comment (i1i1) [inline src/summary.rs:152]: Use `dirs::cache_dir` here

File: src/summary.rs (hunk 2 above)
Comment (ozgunozerk) [reply src/summary.rs:50]: we already have this struct as `Summary`
this struct was intermediate for updating the summary.
After our discussion, it probably won't be used

**Comment by ozgunozerk** [comment]
Reworking, will re-open when I'm done, don't wanna notify with push messages

**Pull request closed without merging**